truffle/src/main/java/org/jruby/truffle/runtime/rope/Rope.java

```diff
@@ -93,4 +106,34 @@ public int realSize() {
     public int getRealSize() {
         return realSize();
     }
+
+    @Override
+    public int hashCode() {
+        if (hashCode == 0) {
+            hashCode = RopeOperations.hashForRange(this, 1, 0, byteLength);
+        }
+
+        return hashCode;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) {
+            return true;
+        }
+
+        if (o instanceof Rope) {
+            final Rope other = (Rope) o;
+
+            if ((hashCode != 0) && (other.hashCode != 0) && (hashCode != other.hashCode)) {
+                return false;
+            }
+
+            // TODO (nirvdrum 21-Jan-16): We really should be taking the encoding into account here. We're currenly not because it breaks the symbol table.
```

Comment:
`String#==` is an odd one because `'abc'.encode('US-ASCII') == 'abc'.encode('UTF-8') => true`. So it can't be a direct encoding comparison anyway.